src/PSR7/Validators/BodyValidator/BodyValidator.php

```diff
@@ -42,38 +49,50 @@ public function validate(OperationAddress $addr, MessageInterface $message) : vo
             return;
         }
 
-        $contentTypes = $message->getHeader(self::HEADER_CONTENT_TYPE);
-        if (! $contentTypes) {
+        // Detect ContentType of the message
+        $contentType = $this->messageContentType($message);
+        if (! $contentType) {
             throw InvalidHeaders::becauseOfMissingRequiredHeader(self::HEADER_CONTENT_TYPE, $addr);
         }
-        $contentType = $contentTypes[0]; // use the first value
-
-        // As per https://tools.ietf.org/html/rfc7231#section-3.1.1.5 and https://tools.ietf.org/html/rfc7231#section-3.1.1.1
-        // ContentType can contain multiple statements (type/subtype + parameters), ie: 'multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__'
-        // OpenAPI Spec only defines the first part of the header value (type/subtype)
-        // Other parameters SHOULD be skipped
-        $contentType = strtok($contentType, ';');
 
         // does the response contain one of described media types?
         if (! isset($mediaTypeSpecs[$contentType])) {
             throw InvalidHeaders::becauseContentTypeIsNotExpected($contentType, $addr);
         }
 
-        // ok looks good, now apply validation
-        $body = (string) $message->getBody();
+        // detect the schema for the media type
+        $schema = $mediaTypeSpecs[$contentType]->schema;
+        if (! $schema) {
+            // no schema means no validation
+            // note: schema is REQUIRED to define the input parameters to the operation when using multipart content
+            return;
+        }
 
-        if (preg_match('#^application/.*json$#', $contentType)) {
-            $body = json_decode($body, true);
-            if (json_last_error() !== JSON_ERROR_NONE) {
-                throw InvalidBody::becauseBodyIsNotValidJson(json_last_error_msg(), $addr);
-            }
+        // Validate message body
+        if (preg_match('#^multipart/.*#', $contentType)) {
+            $this->validateMultipart($addr, $message, $mediaTypeSpecs, $contentType);
```

Comment:
I hope these three could be separate validators i.e

```php
(new MultipartBodyValidator($mediaTypeSpecs, $contentType))->validate($addr, $message)
```

would perfect fit existing `ValidatorInterface` and would be much reusable than trait which us currently used once (I think just make code a bit cleaner)

Reply:
Great idea! Let me try to do that.